File: tools/buildgen/town_hash.py
"""Deterministic, parity-safe hash primitives for seed-derived town geometry.

Every seed-derived geometry parameter (perimeter vocabulary selection, grid
jitter, modifier presence) routes through this module so the Python planner and
the Java realizer reproduce identical parameters from the same ``(seed, tag)``
pair without sharing an RNG stream. See
``src/main/java/com/example/myvillage/town/TownHash.java`` for the bit-identical
mirror.

The primitives are pure-integer (no float) and platform-independent:

- ``hash64(seed, tag)`` — splitmix64-style finalizer over a tagged FNV-1a hash
  of the seed bytes. Returns a non-negative Python int in the unsigned 64-bit
  range; the Java mirror returns the equivalent ``long`` interpreted as
  unsigned via ``Long.rotateRight`` / bit ops over the signed 64-bit value.
- ``range64(seed, tag, lo, hi)`` — inclusive integer range ``[lo, hi]`` derived
  from ``hash64``. ``lo`` and ``hi`` MUST be ints with ``lo <= hi``.
- ``pick(seed, tag, options)`` — deterministic selection from a non-empty
  sequence by ``hash64(seed, tag) % len(options)``.

Design notes (parity-critical):

- The finalizer is splitmix64 (``z = (z + GOLDEN) ; z = (z ^ (z >> 30)) * M1 ;
  z = (z ^ (z >> 27)) * M2 ; z ^= z >> 31``), operating on the unsigned 64-bit
  value. Both ends mask to 64 bits after every arithmetic step.
- The pre-mixer is FNV-1a over the UTF-8 bytes of ``"seed=<n>;tag=<s>"`` so
  different tags with identical seed bytes hash to unrelated outputs. The
  encoding is stable across Python/Java (ASCII seed decimal + tag bytes).
"""

from __future__ import annotations

from typing import Sequence, TypeVar

# splitmix64 constants (Knuth / Steele & Marsaglia).
_SPLITMIX_GOLDEN = 0x9E3779B97F4A7C15
_SPLITMIX_M1 = 0xBF58476D1CE4E5B9
_SPLITMIX_M2 = 0x94D049BB133111EB
_MASK64 = (1 << 64) - 1

T = TypeVar("T")


def _fnv1a_tagged(seed: int, tag: str) -> int:
    """64-bit FNV-1a over the bytes of ``f"seed={seed_u};tag={tag}"``.

    ``seed`` is masked to its unsigned 64-bit representation before encoding so
    a negative Java ``long`` and the corresponding Python int produce identical
    bytes (``Long.toUnsignedString`` / ``int & MASK`` agree on the decimal
    digits). The byte encoding is exactly what the Java mirror constructs, so
    the intermediate hash is bit-identical across ends.
    """
    seed_u = int(seed) & _MASK64
    h = 0xCBF29CE484222325  # FNV-1a 64-bit offset basis
    payload = f"seed={seed_u};tag={tag}".encode("utf-8")
    for b in payload:
        h ^= b
        h = (h * 0x100000001B3) & _MASK64  # FNV-1a 64-bit prime
    return h & _MASK64


def _splitmix64(z: int) -> int:
    """splitmix64 finalizer over an unsigned 64-bit input -> unsigned 64-bit."""
    z = (z + _SPLITMIX_GOLDEN) & _MASK64
    z = (z ^ (z >> 30)) & _MASK64
    z = (z * _SPLITMIX_M1) & _MASK64
    z = (z ^ (z >> 27)) & _MASK64
    z = (z * _SPLITMIX_M2) & _MASK64
    return (z ^ (z >> 31)) & _MASK64


def hash64(seed: int, tag: str) -> int:
    """Deterministic unsigned 64-bit hash of ``(seed, tag)``.

    Bit-identical to ``TownHash.hash64`` on the Java side (the Java mirror
    returns the same bit pattern interpreted through a signed ``long``).
    """
    return _splitmix64(_fnv1a_tagged(int(seed), str(tag)))


def range64(seed: int, tag: str, lo: int, hi: int) -> int:
    """Inclusive deterministic integer in ``[lo, hi]`` derived from ``hash64``.

    ``lo`` and ``hi`` MUST be integers with ``lo <= hi``. Uses integer-only
    arithmetic so the Java mirror reproduces it bit-for-bit (Java computes the
    same modulus over the unsigned-interpreted 64-bit value).
    """
    if lo > hi:
        raise ValueError(f"range64: lo ({lo}) > hi ({hi})")
    span = hi - lo
    return lo + (hash64(seed, tag) % (span + 1))


def pick(seed: int, tag: str, options: Sequence[T]) -> T:
    """Deterministic selection from a non-empty ``options`` sequence."""
    if not options:
        raise ValueError("pick: options must be non-empty")
    return options[hash64(seed, tag) % len(options)]


__all__ = ["hash64", "range64", "pick"]

File: src/main/java/com/example/myvillage/town/TownHash.java
package com.example.myvillage.town;

/**
 * Deterministic, parity-safe hash primitives — bit-identical mirror of
 * {@code tools/buildgen/town_hash.py}. Every seed-derived town-geometry
 * parameter (perimeter vocabulary selection, grid jitter, modifier presence)
 * routes through this class so the Java realizer reproduces the Python
 * planner's parameters from the same {@code (seed, tag)} pair without sharing
 * an RNG stream.
 *
 * <p>The primitives are pure-integer (no float). All right shifts use
 * {@code >>>} (logical) because the Python side masks every step to a
 * non-negative 64-bit value; signed vs unsigned interpretation of the
 * intermediate {@code long} does not affect the low 64 bits after add / mul /
 * shift / xor, so the bit pattern is identical across ends.
 */
public final class TownHash {
    // splitmix64 constants (Knuth / Steele & Marsaglia).
    private static final long SPLITMIX_GOLDEN = 0x9E3779B97F4A7C15L;
    private static final long SPLITMIX_M1 = 0xBF58476D1CE4E5B9L;
    private static final long SPLITMIX_M2 = 0x94D049BB133111EBL;
    private static final long FNV_OFFSET = 0xCBF29CE484222325L;  // FNV-1a basis
    private static final long FNV_PRIME = 0x100000001B3L;

    private TownHash() {
    }

    /**
     * 64-bit FNV-1a over the UTF-8 bytes of {@code "seed=<u>;</tag>"} where
     * {@code <u>} is the unsigned decimal of {@code seed}. Matches the Python
     * {@code _fnv1a_tagged} byte-for-byte.
     */
    private static long fnv1aTagged(long seed, String tag) {
        String payload = "seed=" + Long.toUnsignedString(seed) + ";tag=" + tag;
        byte[] bytes = payload.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        long h = FNV_OFFSET;
        for (byte b : bytes) {
            h ^= (b & 0xFFL);
            h *= FNV_PRIME;  // wraps at 64 bits, matching Python & MASK64
        }
        return h;
    }

    /** splitmix64 finalizer over a 64-bit input (unsigned interpretation). */
    private static long splitmix64(long z) {
        z = z + SPLITMIX_GOLDEN;
        z = (z ^ (z >>> 30)) * SPLITMIX_M1;
        z = (z ^ (z >>> 27)) * SPLITMIX_M2;
        return z ^ (z >>> 31);
    }

    /**
     * Deterministic 64-bit hash of {@code (seed, tag)}. The returned
     * {@code long} carries the same bit pattern as the Python {@code hash64}
     * unsigned integer (compare via {@link Long#compareUnsigned} or bit
     * patterns, not signed ordering).
     */
    public static long hash64(long seed, String tag) {
        return splitmix64(fnv1aTagged(seed, tag));
    }

    /**
     * Inclusive deterministic integer in {@code [lo, hi]} derived from
     * {@code hash64}. Uses unsigned modulus so the Java value matches the
     * Python {@code lo + (hash64 % (span + 1))} bit-for-bit.
     */
    public static int range64(long seed, String tag, int lo, int hi) {
        if (lo > hi) {
            throw new IllegalArgumentException("range64: lo (" + lo + ") > hi (" + hi + ")");
        }
        int span = hi - lo;
        long h = hash64(seed, tag);
        // Unsigned remainder: hash64 is interpreted unsigned on the Python side.
        return lo + (int) Long.remainderUnsigned(h, (span + 1L));
    }

    /**
     * Deterministic selection from a non-empty array by unsigned modulus.
     * Matches Python {@code pick} bit-for-bit.
     */
    public static <T> T pick(long seed, String tag, T[] options) {
        if (options.length == 0) {
            throw new IllegalArgumentException("pick: options must be non-empty");
        }
        return options[(int) Long.remainderUnsigned(hash64(seed, tag), options.length)];
    }
}
